homeassistant/components/http/auth.py

```diff
@@ -69,30 +167,25 @@ def setup_auth(app, trusted_networks, api_password):
               await async_validate_signed_request(request)):
             authenticated = True
 
-        elif (api_password and HTTP_HEADER_HA_AUTH in request.headers and
-              hmac.compare_digest(
-                  api_password.encode('utf-8'),
-                  request.headers[HTTP_HEADER_HA_AUTH].encode('utf-8'))):
-            # A valid auth header has been set
-            authenticated = True
-            request['hass_user'] = await legacy_api_password.async_get_user(
-                app['hass'])
+        elif support_legacy and HTTP_HEADER_HA_AUTH in request.headers:
+            user = await legacy_api_password.async_validate_password(
+                hass, request.headers[HTTP_HEADER_HA_AUTH])
+            if user is not None:
```

Comment:
The downside of having this `if` be inside the `elif`, is that if this method does not match, we won't try the other methods.

Reply:
I'm not saying this is a bad thing perse, but it's a change in behavior.

Reply:
The new walrus operator would have helped here 👍 https://medium.com/hultner/try-out-walrus-operator-in-python-3-8-d030ce0ce601

Reply:
Oh, that is another point we have those separate `validate_` method

Reply:
wrap each  single validation into a function, now the middleware function itself looks better.